### DISCUSSION TITLE
resolve_fks in DeferredReflection

### DISCUSSION THREAD

**Discussion by Dartagnanpl** [sqlalchemy/sqlalchemy#7722] [Usage Questions] [answered]
Hello,
I believe that the option **reflect_fks** available for tables since 1.3 does not work with **DeferredReflection** (or at least I see multiple calls to ALL_CONSTRAINTS when reflecting my Oracle database). This is a feature that I'd like to use a lot, since the database I'm connecting to is very large and its full contraint reflection takes a lot of time. Is there a way to get around that? Or maybe it is planned in next versions of SQLAlchemy?

**Comment by zzzeek** [comment, marked as answer]
hi there -

I *think* you might be referring to the parameter "resolve_fks" on Table, and you are looking to reduce FK lookups when using DeferredReflection.   So yes it looks like this particular parameter isn't passed through.   You can override this as follows:

```py
class MyBase(DeferredReflection, Base):
    __abstract__ = True

    @classmethod
    def _reflect_table(cls, table, inspector):
        Table(
            table.name,
            table.metadata,
            resolve_fks=False
            extend_existing=True,
            autoload_replace=False,
            autoload_with=inspector,
            schema=table.schema,
        )
```


DeferredReflection is not used that often, most people likely use [automap](https://docs.sqlalchemy.org/en/14/orm/extensions/automap.html?highlight=automap#module-sqlalchemy.ext.automap) which does support all reflection options including resolve_fks.

**Comment by CaselIT** [comment]
> since the database I'm connecting to is very large and its full contraint reflection

since we are working on improving the reflections, could you share some statistics regarding schema of the db so that we have a more significant case to test against?
mainly number of tables, columns, and constraints. Or if the schema is public it would be great a link to the sql!

**Comment by Dartagnanpl** [comment]
Thank you very much! This tiny piece of code resulted in a decrease in the length of reflection from 15-20 minutes to 20 seconds!

Yes, I can provide any statistics you wish since I think this is a great project and I would be glad to contribute in any way. 

My case is quite specific: we run a trading company that uses a single oracle distance in its centre and separate oracle instances in each of 30 stores. We use commercial ERP software, but I want to use SQL Alchemy to provide some functionality the ERP software does not (that is why I do not need nor want to reflect full database). Unfortunately, I can't publish the DB.

The "central" database has a very similar schema to stores' databases, but there are some differences regarding the tables. That's why I use deferred reflection - in order to avoid code repetition I created classes representing common tables (the majority of all) and some classes which are specific for stores and for the central database. Some of the constraints are also specific to this "type" of the database, that's another reason for deferred reflection.

Most of the tables are irrelevant for me - out of 5714 tables in the schema I want to reflect only 15. So, the statistics for the central database (in the schema I reflect):
`select count(*) from user_tables; -- 5714`
`select count(*) from user_tab_columns; -- 204 739`
`select count(*) from user_constraints; -- 53 144`

and for stores' database:
`select count(*) from user_tables; -- 3868`
`select count(*) from user_tab_columns; -- 130 845`
`select count(*) from user_constraints; -- 36 482`

**Reply by CaselIT** [reply]
thanks for the data point. That's a quite large db!
I'll check if the example script the generated dbs are similar in composition

> I want to reflect only 15

I think you could specify only the names you are interested in and reflect those

**Reply by Dartagnanpl** [reply]
And I do, but some of those tables are associated with a couple of hundreds of constraints, hence my original question.

**Reply by CaselIT** [reply]
current work in progress version looks promising.
about 10-20s (following runs look better) to reflect 2500 tables on a oracle xe running on a local docker instances

```
Meta info {'tables': 2500, 'columns': 81545, 'index': 7852, 'check_con': 834, 'foreign_keys_con': 4044, 'unique_con': 3639, 'identity': 240}
{'populate-meta': 4.412026643753052,
 'create-tables': 205.1563322544098,
 'get_table_names': 0.9693582057952881,
 'get_multi_table_columns': 7.873682260513306,
 'get_multi_table_options': 0.979778528213501,
 'get_multi_table_pk_constraint': 3.2456064224243164,
 'get_multi_table_comment': 0.37853336334228516,
 'get_multi_table_check_constraints': 1.7195069789886475,
 'get_multi_table_indexes': 5.264635324478149,
 'get_multi_table_foreign_keys': 2.542823553085327,
 'get_multi_table_unique_constraints': 1.816206455230713}
```
